Address oversized allocs across kafka API and schema registry

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Previously, many API types across Kafka API, HTTP Proxy, and Schema Registry used `std::vector` to hold request contents or results. Dues to `std::vectors` contiguous memory guarantees, this could lead to oversized allocations under certain usage patterns.

This PR introduces a number of scale tests to reproduce these overallocs and also makes an effort to replace `std::vector` with `chunked_vector` wherever possible/expedient. Some element types were left out to account for copyability requirements enforced by e.g. seastar's continuation helper functions.

Also moves some array serialization in SR to a streaming model, also to avoid oversized allocs.



## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v23.3.x
- [ ] v23.2.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman commented:
/dt

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/8a349d385ebed622f4bca1d91d9a8633243014d4..8513731885b4414f976ea950d12c5e62eb0635f8) remove dead code (triggering UAM in clang-tidy)

oleiman commented:
/dt

oleiman commented:
/dt

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/17b850a0fef7148663f0072bf7a37d24fe0cb95f..abfb3db08c035fadddb09ccf73a0ed282d32b4f3) - cruft and added vector overrides for txn_offset_commit

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/47839#018ee581-3a7a-4eaa-ad08-10ca8a594c7a:
```
"rptest.tests.write_caching_fi_test.WriteCachingFailureInjectionTest.test_unavoidable_data_loss"
```
<!--ccf83de5-b7de-43aa-aac0-88f87c053930-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/47839#018ee581-3a7e-4950-a275-358c8e4e2e0b
<!--b099e386-fac2-475f-bd84-456511c5d03b-->

oleiman commented:
> Sorry does this do a bulk search & replace of `std::vector` with chunked_vector?

It does not. The order of operations was:

- change the default container type in generator.py `std::vector -> chunked_vector`
- fix compiler errors case by case
   - if the error is hard (e.g. container captured by a copyable lambda) or the list is always short (e.g. brokers), override with `std::vector`

There should not be any instances of "hey, this looks like it could/should be a `chunked_vector`". Or rather, if any such change looks extraneous, please flag and we will remove it.

michael-redpanda commented on src/v/kafka/protocol/schemata/generator.py:
```diff
@@ -473,6 +473,32 @@
     'creatable_topic_result': 'chunked_vector',
     'describe_configs_resource_result': 'chunked_vector',
     'describe_configs_result': 'chunked_vector',
+    'incremental_alterable_config': 'chunked_vector',
```
we should probably remove the 'chunked_vector' overrides in the generator.py file

oleiman replied:
Done

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/abfb3db08c035fadddb09ccf73a0ed282d32b4f3..3947398ad639db22aed7573ce41a56116bd661a8) remove redundant chunked_vector overrides from generator.py

BenPope commented on src/v/kafka/server/handlers/list_offsets.cc:
```diff
@@ -318,16 +320,21 @@ list_offsets_handler::handle(request_context ctx, ss::smp_service_group ssg) {
+        return ss::when_all_succeed(topics.begin(), topics.end())
           .then([&octx](std::vector<list_offset_topic_response> topics) {
-              octx.response.data.topics = std::move(topics);
+              octx.response.data.topics.reserve(topics.size());
+              std::move(
+                topics.begin(),
+                topics.end(),
+                std::back_inserter(octx.response.data.topics));
```
I think it would be clearer like this, since the `unauthorized_topics` is roughly the same type:
```suggestion
              octx.response.data.topics = {
                std::make_move_iterator(topics.begin()),
                std::make_move_iterator(topics.end())};

oleiman replied:
yeah good point

BenPope commented on src/v/kafka/server/handlers/metadata.cc:
```diff
@@ -162,7 +163,8 @@ metadata_response::topic make_topic_response_from_topic_metadata(
         }
         p.replica_nodes = std::move(replicas);
-        p.isr_nodes = p.replica_nodes;
+        p.isr_nodes.reserve(p.replica_nodes.size());
+        absl::c_copy(p.replica_nodes, std::back_inserter(p.isr_nodes));
```
Not needed

BenPope commented on src/v/kafka/server/tests/fetch_session_test.cc:
```diff
@@ -175,8 +176,14 @@ FIXTURE_TEST(test_session_operations, fixture) {
     {
         // Remove and forget about the first partition.
-        req.data.topics[0].fetch_partitions.erase(
+        auto parts = req.data.topics[0].fetch_partitions.copy();
+        parts.erase_to_end(
           std::next(req.data.topics[0].fetch_partitions.begin()));
+        std::move(
+          std::next(req.data.topics[0].fetch_partitions.begin(), 2),
+          req.data.topics[0].fetch_partitions.end(),
+          std::back_inserter(parts));
+        req.data.topics[0].fetch_partitions = std::move(parts);
```
I don't think this does what you want it to.

oleiman replied:
yikes

oleiman commented on src/v/kafka/client/client.cc:
```diff
@@ -512,7 +530,7 @@ ss::future<offset_commit_response> client::consumer_offset_commit(
   std::vector<offset_commit_request_topic> topics) {
     return get_consumer(g_id, name)
-      .then([this, topics{std::move(topics)}](shared_consumer_t c) mutable {
+      .then([this, topics = std::move(topics)](shared_consumer_t c) mutable {
```
indeed

oleiman commented on src/v/kafka/server/handlers/alter_configs.cc:
```diff
@@ -371,14 +371,25 @@ ss::future<response_ptr> alter_configs_handler::handle(
           alter_configs_resource>(
-          std::move(groupped), std::move(unauthorized_responsens));
+          std::move(groupped),
+          std::vector<alter_configs_resource_response>{
+            std::move_iterator(unauthorized_responsens.begin()),
+            std::move_iterator(unauthorized_responsens.end())});
+
+        std::vector<chunked_vector<alter_configs_resource_response>>
+          responses_cv;
+        responses_cv.reserve(responses.size());
+        for (const auto& v : responses) {
+            responses_cv.emplace_back(
+              std::move_iterator(v.begin()), std::move_iterator(v.end()));
+        }
         co_return co_await ctx.respond(
           assemble_alter_config_response<
             alter_configs_response,
-            alter_configs_resource_response>(std::move(responses)));
+            alter_configs_resource_response>(std::move(responses_cv)));
```
yeah, skipped due to time pressure. moot point; I'll take a look.

oleiman commented on src/v/kafka/server/handlers/api_versions.cc:
```diff
@@ -106,11 +106,14 @@ api_versions_response api_versions_handler::handle_raw(request_context& ctx) {
-            r.data.api_keys = supported_apis.base;
+            absl::c_copy(
+              supported_apis.base, std::back_inserter(r.data.api_keys));
         } else if (!ctx.are_transactions_enabled()) {
-            r.data.api_keys = supported_apis.idempotence;
+            absl::c_copy(
+              supported_apis.idempotence, std::back_inserter(r.data.api_keys));
         } else {
-            r.data.api_keys = supported_apis.transactions;
+            absl::c_copy(
+              supported_apis.transactions, std::back_inserter(r.data.api_keys));
```
ugh, though I caught all of these

oleiman commented on src/v/kafka/server/handlers/incremental_alter_configs.cc:
```diff
@@ -461,14 +461,23 @@ static ss::future<std::vector<resp_resource_t>> alter_broker_configuartion(
             continue;
         }
 
+        req_resource_t resource_c{
+          .resource_type = resource.resource_type,
+          .resource_name = resource.resource_name,
+          .unknown_tags = resource.unknown_tags,
+        };
+        resource_c.configs.reserve(resource.configs.size());
+        absl::c_copy(resource.configs, std::back_inserter(resource_c.configs));
```
>brittle

very 

>wonder if we should generate copy constructors

Discussed this at some point. Didn't seem feasible given the old schedule. We still need to get something out, but I may spend an hour today to see how tricky.

>maybe assert they're equal

Better than nothing!

oleiman commented on src/v/kafka/server/handlers/list_offsets.cc:
```diff
@@ -318,16 +320,21 @@ list_offsets_handler::handle(request_context ctx, ss::smp_service_group ssg) {
       std::make_move_iterator(request.data.topics.end()));
 
-    request.data.topics.erase(unauthorized_it, request.data.topics.end());
+    request.data.topics.pop_back_n(
+      std::distance(unauthorized_it, request.data.topics.end()));
```
humbug

oleiman commented on src/v/kafka/client/consumer.cc:
```diff
@@ -365,9 +360,10 @@ ss::future<describe_groups_response> consumer::describe_group() {
 consumer::offset_fetch(std::vector<offset_fetch_request_topic> topics) {
     refresh_inactivity_timer();
-    auto req_builder = [topics{std::move(topics)}, group_id{_group_id}] {
+    auto req_builder = [topics{std::move(topics)},
+                        group_id{_group_id}]() mutable {
```
this one's moot, anyway; vector param. missed on second pass review.

oleiman commented on src/v/kafka/client/consumer.cc:
```diff
@@ -398,12 +391,13 @@ consumer::offset_commit(std::vector<offset_commit_request_topic> topics) {
         }
     }
-    auto req_builder = [me{shared_from_this()}, topics{std::move(topics)}]() {
+    auto req_builder = [me{shared_from_this()},
+                        topics{std::move(topics)}]() mutable {
```
same as above

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/3947398ad639db22aed7573ce41a56116bd661a8..5465c0fde2d2b79271616e4ae2bff2ac24fde8d1) contents:
- removed a bunch more junk per CR comments
- added a const constraint to `consumer::req_res`

BenPope commented on src/v/container/include/container/lw_shared_container.h:
```diff
@@ -0,0 +1,32 @@
+template<typename C>
+class lw_shared_container {
+public:
+    using iterator = C::iterator;
```
since it's shared, is it worth making it const?

oleiman replied:
Compiles as suggested, so I'm going to say :shipit:

BenPope commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -238,6 +240,41 @@ get_schemas_ids_id(server::request_t rq, server::reply_t rp) {
 }
 
+struct schema_version_json : public ss::json::json_base {
+    ss::json::json_element<ss::sstring> subject;
```
It's a shame this can't be a `std::string_view`, but I'm not sure it's worth fixing for now.

oleiman replied:
yeah...this usage pattern is endemic to types derived from `ss::json::json_base` (for our purposes, mostly generated code) fwiw. 

I'd probably prefer a move-assignable `json_element` anyway.

BenPope commented on tests/rptest/scale_tests/many_topics_test.py:
```diff
@@ -1319,3 +1329,167 @@ def test_many_topics_config(self):
+            f"{r.status_code} {verb} hostname={hostname} {path} {kwargs}")
+
+        return r
+
+    def _post_subjects_subject_versions(self,
+                                        subject,
+                                        data,
+                                        headers=HTTP_POST_HEADERS,
+                                        **kwargs):
+        return self._request("POST",
+                             f"subjects/{subject}/versions",
+                             headers=headers,
+                             data=data,
+                             **kwargs)
+
+    def _get_subjects(self, deleted=False, headers=HTTP_GET_HEADERS, **kwargs):
+        return self._request("GET",
+                             f"subjects{'?deleted=true' if deleted else ''}",
+                             headers=headers,
+                             **kwargs)
+
+    def _get_schemas_ids_id_versions(self,
+                                     id,
+                                     headers=HTTP_GET_HEADERS,
+                                     **kwargs):
+        return self._request("GET",
+                             f"schemas/ids/{id}/versions",
+                             headers=headers,
+                             **kwargs)
+
+    def _get_schemas_ids_id_subjects(self,
+                                     id,
+                                     deleted=False,
+                                     headers=HTTP_GET_HEADERS,
+                                     **kwargs):
+        return self._request(
+            "GET",
+            f"schemas/ids/{id}/subjects{'?deleted=true' if deleted else ''}",
+            headers=headers,
+            **kwargs)
```
Is any of this fundamentally different from `SchemaRegistryEndpoints`? Is it worth extracting that a bit so it can be shared?

oleiman replied:
Looks reasonable at first glance, but I may defer to @michael-redpanda's intuition.

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/5465c0fde2d2b79271616e4ae2bff2ac24fde8d1..2675be28456115cf8e49364c37c5a634ad29ad77) SR related CR comments

dotnwat reviewed (commented):
> kafka: Change default container to chunked_vector for generated code

why do this in one commit?

oleiman commented:
> > kafka: Change default container to chunked_vector for generated code
> 
> 
> 
> why do this in one commit?

Bisectability. Once we change the default container, the whole diff is required to fix the build. Agree the end result is not great.

michael-redpanda commented:
> > > kafka: Change default container to chunked_vector for generated code
> > 
> > 
> > why do this in one commit?
> 
> Bisectability. Once we change the default container, the whole diff is required to fix the build. Agree the end result is not great.

We did discuss doing it "in reverse" where we would attempt to identify every structure and override to use `chunked_vector`.  The time constraints _at the time_ pushed us to do it wholesale and then override back to `std::vector` in certain cases.

BenPope reviewed (approved):
Would prefer if the refactoring for the Schema Registry request bits was done in this PR.

BenPope commented on src/v/kafka/server/handlers/alter_partition_reassignments.cc:
```diff
@@ -375,9 +376,10 @@ static ss::future<response_ptr> do_handle(alter_op_context& octx) {
           return ssx::parallel_transform(
-            octx.request.data.topics.begin(),
-            octx.request.data.topics.end(),
+            std::make_move_iterator(octx.request.data.topics.begin()),
+            std::make_move_iterator(octx.request.data.topics.end()),
```
> is this safe to do?

I believe so.

ballard26 commented on src/v/kafka/client/client.cc:
```diff
@@ -289,10 +289,13 @@ ss::future<produce_response> client::produce_records(
             std::move(*p.records->adapter.batch));
       });
 
+    chunked_vector<topic_produce_response> responses_cv;
```
nit: Is there something preventing us from using list-initialization for the chunked_vector like we did previously with the vector?

ballard26 replied:
Right you are `std::initializer_list` only returns constant iterators. Dunno how I missed it before. Definitely something we should avoid using if possible.

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/2675be28456115cf8e49364c37c5a634ad29ad77..313d2084299e33225694ebf0045cfbfdb0f550f2) - remove struct size assert (differs between debug/release). 

poking at it a bit locally to see whether there's a clean way to express the same intent.

ballard26 commented on tests/rptest/scale_tests/many_topics_test.py:
```diff
@@ -36,6 +37,13 @@
+}
+
 
 class ManyTopicsTest(RedpandaTest):
```
Did we decide that testing server-side validation wasn't needed?

ballard26 reviewed (approved):
LGTM from a perf perspective.

dotnwat commented:
> > > > kafka: Change default container to chunked_vector for generated code
> > > 
> > > 
> > > why do this in one commit?
> > 
> > 
> > Bisectability. Once we change the default container, the whole diff is required to fix the build. Agree the end result is not great.
> 
> We did discuss doing it "in reverse" where we would attempt to identify every structure and override to use `chunked_vector`. The time constraints _at the time_ pushed us to do it wholesale and then override back to `std::vector` in certain cases.

The question wasn't in regard to how many are changed, it was that they all seem to be changed in a single commit. If one of the data types in that large commit changes, does it mean that all of the other changes flow from that? Seems like there should be a way to contain the fall out from a change in order to build up smaller changes.

oleiman commented:
> > > > > kafka: Change default container to chunked_vector for generated code
> > > > 
> > > > 
> > > > why do this in one commit?
> > > 
> > > 
> > > Bisectability. Once we change the default container, the whole diff is required to fix the build. Agree the end result is not great.
> > 
> > 
> > We did discuss doing it "in reverse" where we would attempt to identify every structure and override to use `chunked_vector`. The time constraints _at the time_ pushed us to do it wholesale and then override back to `std::vector` in certain cases.
> 
> The question wasn't in regard to how many are changed, it was that they all seem to be changed in a single commit. If one of the data types in that large commit changes, does it mean that all of the other changes flow from that? Seems like there should be a way to contain the fall out from a change in order to build up smaller changes.

Not 100% sure I'm answering your question @dotnwat, but yeah, all of the changes in that big commit flow directly from changing the default container type in `generator.py`. And at that mostly on account of `chunked_vector` being non-copyable.

oleiman commented:
CI failures are https://github.com/redpanda-data/redpanda/issues/18033

vbotbuildovich commented:
/backport v24.1.x

vbotbuildovich commented:
/backport v23.3.x

vbotbuildovich commented:
Failed to create a backport PR to v23.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-17876-v23.3.x-203 remotes/upstream/v23.3.x
git cherry-pick -x 55436c6ed4a246c59ca69dcaa3287f31090765e5 0ef53aada284f8de69040f10955252e375b295f1 62e5aa94155579804018d2b600668f76c3fa4c30 b2b57325b13e36fc40969adcda18b1bf7863d8b8 3aadeecc7c4fab53b06d27b6dc7bc590091f293c 2004a99ee7ebc9923260bffb1da2fc265d147f17 313d2084299e33225694ebf0045cfbfdb0f550f2
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/8817329719).

oleiman commented:
> Right, but isn't an option to not change the _default_ -- rather change them one-by-one?

Yeah, that's the alternative Mike was describing above. At the time we were under time pressure and this was the most expedient path.

dotnwat commented:
> > Right, but isn't an option to not change the _default_ -- rather change them one-by-one?
> 
> Yeah, that's the alternative Mike was describing above. At the time we were under time pressure and this was the most expedient path.

👍
